numpy/core/src/common/ufunc_override.c

```diff
@@ -93,8 +93,16 @@ PyUFuncOverride_GetOutObjects(PyObject *kwds, PyObject **out_kwd_obj, PyObject *
         return 0;
     }
     if (PyTuple_CheckExact(*out_kwd_obj)) {
-        *out_objs = PySequence_Fast_ITEMS(*out_kwd_obj);
-        return PySequence_Fast_GET_SIZE(*out_kwd_obj);
+        PyObject *seq = PySequence_Fast(*out_kwd_obj, "cannot convert");
```

Comment:
So on pypy, exact tuples are not also always fast sequences?

Reply:
They need conversion to a "fast strategy tuple". PyPy will convert a tuple of like items (all int) to an array of unboxed values. In order to support fast access, PyPy must box the values into an array of `PyObject*`, and build a CPython compatible PyTupleObject. Since this is expensive, it is only done when calling `PySequence_Fast()`

Reply:
Note that the PyPy implementation of `PySequence_FAST` contradicts the CPython docs (emphasis mine):

> `PyObject* PySequence_Fast(PyObject *o, const char *m)`
> _Return value: New reference._
>
> Return the sequence or iterable o as a list, **unless it is already a tuple** or list, **in which case o is returned**. Use `PySequence_Fast_GET_ITEM()` to access the members of the result. Returns NULL on failure. If the object is not a sequence or iterable, raises `TypeError` with m as the message text.

So I think we need to emphasize in the comment that the documentation for `PySequence_Fast` does not apply to PyPy.

It might be worth contacting the CPython devs and recommending they use weaker language in their documentation, to not make this guarantee.

Reply:
see python/cpython#11140